### DISCUSSION TITLE
Interleaved data

### DISCUSSION THREAD

**Discussion by Yay295** [python-pillow/Pillow#6673] [Q&A] [answered]
Pillow currently supports "band interleaved by pixel" and "band interleaved by line" image data (to an extent). Should it also support "band sequential"? Supporting only one layout would be the easiest, but if someone wants to use their image data with some other library they might want it to be in a specific layout.

https://desktop.arcgis.com/en/arcmap/latest/manage-data/raster-and-images/bil-bip-and-bsq-raster-files.htm

**Comment by radarhere** [comment]
I think Pillow already does. Here is a snippet from SgiImagePlugin, where one channel is loaded after another into the final image.

https://github.com/python-pillow/Pillow/blob/c924fd84a39f3ee9aa9b96d8dec58ec50fdca9d6/src/PIL/SgiImagePlugin.py#L114-L118

**Reply by Yay295** [reply]
That's before the data gets decoded.

https://github.com/python-pillow/Pillow/blob/c924fd84a39f3ee9aa9b96d8dec58ec50fdca9d6/src/PIL/SgiImagePlugin.py#L200-L216

On line 214 is `self.im.putband(channel.im, band)`. `putband` is C code that can be traced back to here:

https://github.com/python-pillow/Pillow/blob/d72779ac03b73f9f9538c8e7fc9f57c887452e88/src/libImaging/Bands.c#L169-L208

So it is loading the raw data one channel/layer/band at a time, but when the data is decoded it is being stored as "rgbrgbrgb", not "rrrgggbbb".

**Reply by radarhere** [reply]
What would be the advantage to storing the data as rrrgggbbb? If it is so that another library can use Pillow's C code directly, then my response is that Pillow is a Python library that happens to have C code inside it, [not a C library](https://github.com/python-pillow/Pillow/issues/4532#issuecomment-618255798).

**Reply by Yay295** [reply]
Yep, that makes sense. Check my other comment; I was mistaken about Pillow already using multiple data layouts.

**Comment by Yay295** [comment, marked as answer]
The interleaved data I had remembered seeing is actually in the unpacking code:

https://github.com/python-pillow/Pillow/blob/d72779ac03b73f9f9538c8e7fc9f57c887452e88/src/libImaging/Unpack.c#L415-L435

So after the data is unpacked, it is being stored as "band interleaved by pixel", not "band interleaved by line". So this isn't an issue; all image data is being stored the same after decoding/unpacking.
